ocaml/xapi/dbsync.ml

```diff
@@ -47,6 +47,14 @@ let create_host_metrics ~__context =
 
 
 let update_env () =
+
+  (** Allows code in cancel_tasks.ml to clean up any orphaned tasks *)
+  let set_me_as_resident_on ~__context =
+    let task = Context.get_task_id __context in
+    let me_ref = Helpers.get_localhost ~__context in
+    Db.Task.set_resident_on ~__context ~self:task ~value:me_ref;
+  in
+
   Server_helpers.exec_with_new_task "dbsync (update_env)" ~task_in_database:true
```

Comment:
This function sets the `resident_on` field of the new task to `!Xapi_globs.localhost_ref`: https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/taskHelper.ml#L43. However, `Xapi_globs.localhost_ref` is set to the local host ref only later in dbsync: https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/dbsync_slave.ml#L245. Perhaps we can set this ref earlier. For example, set it right immediately where it is defined (https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_globs.ml#L36), rather than making it null.

Reply:
i didn't want to do that because I was worried about breaking other clients, but if you think it wouldn't then that would certainly make sense

Reply:
Oh no, we can't because `Helpers.get_localhost` needs the DB, which isn't immediately available. I think that we could set the ref right above here though.

Reply:
i misunderstood this yesterday, but it makes sense now.

what might be even better is to have `Helpers.get_localhost` cache the host ref as a side effect